enclave/core/sgx/calls.c

```diff
@@ -131,6 +134,51 @@ extern bool oe_disable_debug_malloc_check;
 **==============================================================================
 */
 
+#ifdef OE_WITH_EXPERIMENTAL_EEID
+extern const volatile oe_sgx_enclave_properties_t oe_enclave_properties_sgx;
+extern oe_sgx_enclave_properties_t oe_enclave_properties_sgx_mutable;
+extern volatile int _have_eeid;
+
+static oe_result_t _eeid_patch_memory_sizes()
```

Comment:
Can you comment on why this function is necessary? It wasn't covered in the design document. Looking at `_add_eeid_pages()`, the `props` values for the data page sizes are already updated to the `eeid` specified values before `_add_data_pages()` is invoked, and it looks like the layout at creation should not be overlapping eeid data with heap pages already.

Also, I'm unclear on why moving it after the heap pages makes the eeid safe for later use? The TCS and stack pages are after that so it seems like that would conflict with those data structures?

Reply:
You're right, we made this change after we finished the design docs. The problem is that OE patches some symbols in the base image before it signs it (and so did the EEID code). For the extended image, we want exactly identical patching, so as not to change the base image hash state. The current solution starts the enclave and then immediately patches the EEID symbols after enclave startup. What we ultimately gain from this is that the verifier does not need to have a copy of the base image to re-patch, the hash state is sufficient.

Re moving the EEID: This is currently at the beginning of the heap, so it would get overwritten soon after enclave creation. There are extra pages after the heap that are reserved for this, but we can't save their size or their location without changing the hash state before addition of the memory pages. If you see a way to do that, I'm happy to change this!

Reply:
> Re moving the EEID: This is currently at the beginning of the heap, so it would get overwritten soon after enclave creation.

@wintersteiger I don't understand why the pages need to be added to overlap with the heap. Why not just load them as separate pages? Then you don't need to move them.

Reply:
Actually as @CodeMonkeyLeet `_add_eeid_pages` appears to load them as separate pages. So there should be no overlap with the heap. Maybe the heap offset isn't being update properly?

Reply:
They don't overlap; there are separate heap and EEID pages. However, we can't save the EEID page address/offset in the image (putting that number into the image would change the MRENCLAVE). At the time of entering the enclave, the properties still say heap/stack are zero, so we can't find the end of the heap either. The only reliable address we have is the beginning of the heap, which contains a magic value that tells us that there is EEID to load/move.

Reply:
oe_sgx_enclave_image_info_t stores the relative virtual addresses for different enclave pages. That would be a constant rva for a given enclave.

Adding pages in general will change the MRENCLAVE from the original image anyway, right?

Reply:
`oe_sgx_enclave_image_info_t` saves info about the _base_ image, not about the extended image. It contains the `heap_rva` (beginning of the heap) and `enclave_size` is the base image size (unchanged during enclave creation).  
The critical bit is the hash state before adding any memory pages, which must be unchanged to avoid the need for a copy of the base image that we could re-patch during verification.

Reply:
(`enclave_size` is one of the symbols that are patched later, during enclave startup in `_eeid_patch_memory_sizes`).

Reply:
So let me know if I understand correctly:

The EEID pages are loaded before the heap, which means the rva's in `oe_sgx_enclave_image_info_t` are now incorrect. But modifying the values in `oe_sgx_enclave_image_info_t` would change the pre-data measurement as that struct is measured earlier.

One way I see around this is adding the eeid pages at the very end of the image. Then you can get the eeid pages at enclave_base + enclave_size. Is there a reason that won't work?

Reply:
Yes, that won't work, because `enclave_size` is the size of the base image, which, after extension, points to somewhere in the middle of the data pages. At enclave initialization time, we truly don't know where the memory ends, because we didn't save that information in the image, only on a data page, and the only thing we can rely on is that there is at least one of them (or we segfault).

Reply:
I think this goes back to my question in a previous comment on why the base image size needs to be different from the resulting image size to begin with. Assuming:
- `enclave_size` always reflects expected `secs.size` (max size)
- eeid is written after the end of the data pages
Is there still a concern?

Reply:
Per conversation with @wintersteiger, the remaining issue is that the `oe_enclave_properties_sgx.header.size_settings` (heap page count, stack page count, thread count) are still measured as part of the base image TEXT as a global, and that still needs to be fixed up at runtime because the EEID design requires that to be variable regardless of the base image. 

Not moving the EEID data can still help simplify this though. Consider:
* if `enclave_size` is accurate on base image load and eeid is written after data pages.
* instead of writing just `OE_EEID_MAGIC` into the heap, write the offset to the EEID as well. (i.e. you could define a two field struct)
* in enclave initialization, use the offset to update the mutable enclave properties and zero the first heap page.

Reply:
(Done, see other thread on `_eeid_patch_memory_sizes()`)